client/transaction-pool/src/revalidation.rs

```diff
@@ -131,7 +131,7 @@ impl<Api: ChainApi> RevalidationWorker<Api> {
 
 	fn prepare_batch(&mut self) -> Vec<ExHash<Api>> {
 		let mut queued_exts = Vec::new();
-		let mut left = BACKGROUND_REVALIDATION_BATCH_SIZE;
+		let mut left = std::cmp::max(BACKGROUND_REVALIDATION_BATCH_SIZE, self.members.len() / 4);
```

Comment:
Okay, actually when reading this again, what should this change bring?

And `BACKGROUND_REVALIDATION_BATCH_SIZE` should now be `MIN_BACKGROUND_REVALIDATION_BATCH_SIZE`

Reply:
As I commented, this should dissolve revalidation queue faster after block events 
(when it is most full, since block event is what dumps transactions for revalidation)

Reply:
Did we run into any problems related to transactions not being discarded fast enough?

And blocks only dump transaction for revalidation when we had a Re-org and than one of the retracted blocks need to have a lot of transactions.

I assume we don't revalidate transactions that are pruned through being included in a block?

Reply:
Yep, there are problem with revalidation queue bloating when the chaintip is substantial and new transactions also arrive at the substantial rate

Revalidation dumps everything in the pool (which can be not empty after pruning included transaction, unlike what you probably observe on non-stressed node) to the queue, so it is not only about something retracted. And it is best that most of this revalidated before next block import/production start so that revalidation won't interfere with that.

Reply:
> I assume we don't revalidate transactions that are pruned through being included in a block?

No

Reply:
Did you tested your change? Where did you see this behavior? On Kusama?

I just take another look at the code and randomly increasing the batch size will probably bring us into other problems. The re-validation interval is set to 200ms. This means if the batch size is too big, we will re-validate one batch and directly after this re-validate the next batch. This would run in a loop until some txs are dropped from the pool. This 200ms already makes no sense, why do we need to re-validate multiple times on the same block? The tx status will not change. The re-validation logic should be changed in a more sophisticated way and not just by changing the batch size.

Reply:
Yes, I test it right now under load, and it is much better than before.

Loop is by design, we need constantly revalidate transactions against new block if there are some left and chain increased in length. 

We don't revalidate same transactions "multiple times on the same block", we just try to disperse revalidation of the pool content through the time left to the next block. Tx status can change at block #1001 because it was validated against block #1000 only and we are at #1001 now.

Here I just change it from something stupid like this:

20, 20, 20, 20, 20... 20 - which makes little sense when there are, for example 4000 transactions I the block, only when there are very few transactions.

to something like that:

1000, 750, 500, 350... 20, 0.

Reply:
>  if the batch size is too big, we will re-validate one batch and directly after this re-validate the next batch

I know this is a bit suboptimal, but the good thing is that validation is parallelized and limited to 2 threads, so validation tasks will just queue on those threads if overflows.

Reply:
Where do we spawn 2 threads? I just see that we spawn one future for the re-validation worker.

Reply:
revalidation worker does not validate on it's own, it does it via `ChainApi`, which spawns 2 threads

Reply:
You are right that we spawn the validation on a different pool [here](https://github.com/paritytech/substrate/blob/master/client/transaction-pool/src/api.rs#L83-L123). However, [here](https://github.com/paritytech/substrate/blob/master/client/transaction-pool/src/revalidation.rs#L77) we directly await on the future, which means that we will always only check one transaction at a time. We would need to use a `join_all` to wait for all together to have a validation going on in parallel.

Reply:
addressed those concerns as well @bkchr